test/src/sdk-harness/test_projects/storage_vec/u8/mod.rs

```diff
@@ -0,0 +1,163 @@
+mod utils;
+
+use utils::{
+    setup::get_contract_instance,
+    wrappers::{clear, get, insert, is_empty, len, pop, push, remove, swap_remove},
+};
+
+mod success {
+    use super::*;
+
+    #[tokio::test]
+    async fn can_get_contract_id() {
+        let (_instance, _id) = get_contract_instance().await;
+    }
+
+    #[tokio::test]
+    async fn can_push() {
+        let (instance, _id) = get_contract_instance().await;
+
+        push(&instance, 50).await;
```

Comment:
Each test performs a "post conditional assertion" which checks that the code has the correct value after all of the operations. The issue here is that there are no "pre conditional assertions".
The reason that pre-conditions are required is that there is no guarentee that the code works as expected which could result in a false positive pass.

For example, you could push the value 10 into the vector however if there is no check to make sure that 10 is not already in there then the test will pass with a hidden bug.

The false positive can occur in 3 ways

- The test can be written incorrectly and thus it passes when it should not
- The logic inside the function is incorrect (e.g. it always pushes the value of 10 and the test happens to check if 10 has been pushed)
- The function logic and test may be correct however there is an underlying bug in the VM which makes the test pass when it should not

The solution to this is to first assert the initial condition just before the test (function call), and no earlier, and then right after the function call. This ensures that the state is exactly what you expect it to be and thus the only thing that should change state is the function which is being tested.

Reply:
I cant do preconditional assertions for most of them due to options not being supported, for len and is_empty I could do some preconditional assertions, for the rest of them it seems unfeasible since any expected failures would panic.

I.e. I cant check if 10 is already in there because a get() call would panic in a normal scenario

Reply:
The panics do not need such assertions.
The remainder can use checks on len and capacity, possibly other ways. Nevertheless, preconditions should be added even if we cannot add the correct checks atm because of the limitations of the SDK

Reply:
Suggest leaving some TODOs in there for tests you'd like to add but can't due to current limitations. Doesn't even have to be code, just a description of the tests suffices.